### PR TITLE
node,scylla_node,cluster,scylla_cluster: Add type hints for nodes and clusters

### PR DESCRIPTION
**Pull request by cezarmoise**
- Add type hints for nodes and clusters to help debugging in dtest
- Fix linting / formatting issues

### CODE REVIEW

**Comment by cezarmoise** [comment]
I will do a jenkins run to check for issues

**Comment by cezarmoise** [comment]
> I would recommend add some linter that can identify those, and configure it specific to check those files in CI.
> 
> I think ruff might be use to force adding type notation
> 
> or `mypy` --disallow-untyped-functions

Those tools check all function parameters. For example, with ruff I get that there are about 3000 untyped parameters or function returns. 
I just added hints where nodes and clusters were involved, so it's easier for the editor to navigate trough their methods.

**Comment by cezarmoise** [comment]
> I will do a jenkins run to check for issues

https://jenkins.scylladb.com/view/master/job/scylla-master/job/byo/job/dtest-byo/564/
Ran without issues

**Comment by cezarmoise** [comment]
https://github.com/scylladb/scylla-ccm/commit/df0a17a748a30e5fbd478290a734159b42f42535 rebased to solve conflicts

**Comment by fruch** [comment]
> My only issue is that `typing.*` are deprecated since 3.9, because now you can use inbuilt types for typing instead, e.g. `test: list[str] = []`
> 
> See https://docs.python.org/3/library/typing.html#aliases-to-types-in-collections
> 
> Not sure whaat is our minimal working version, but 3.9 seems old & stable enough to do it the new way. @fruch WDYT?

The minimum is 3.8, we need to support the same versions as the python-driver does, cause ccm is used in it's testing

The typing.* isn't yet deprecated AFAIK, linters are suggesting not to use it.

File: ccmlib/node.py [hunk 1]
```diff
@@ -1125,7 +1127,7 @@ def get_sstablespath(self, output_file=None, datafiles=None, keyspace=None, tabl
         return sstablefiles
 
     def run_sstablerepairedset(self, set_repaired=True, datafiles=None, keyspace=None, column_families=None):
-        cdir = self.get_install_dir()
+        cdir = self.get_install_dir()  # noqa: F841
```
[the remark this answers is not in the document]
Comment (fruch) [reply ccmlib/node.py:1130]: Cause we never linted this code.

This code path also probably isn't in use...

File: ccmlib/cluster.py [hunk 2]
```diff
@@ -246,14 +249,14 @@ def add(self, node, is_seed, data_center=None, rack=None):
 
     # nodes can be provided in multiple notations, determining the cluster topology:
     # 1. int - specifying the number of nodes in a single DC, single RACK cluster
-    # 2. list[int] - specifying the number of nodes in a multi DC, single RACK per DC cluster
+    # 2. List[int] - specifying the number of nodes in a multi DC, single RACK per DC cluster
```
Comment (fruch) [inline ccmlib/cluster.py:252]: No real need to change those comments

File: ccmlib/cluster.py [hunk 3]
```diff
@@ -328,8 +331,8 @@ def populate(self, nodes, debug=False, tokens=None, use_vnodes=False, ipprefix=N
             self._update_config()
         return self
 
-    def new_node(self, i, auto_bootstrap=False, debug=False, initial_token=None, add_node=True, is_seed=True, data_center=None, rack=None):
-        ipformat = self.get_ipformat()
+    def new_node(self, i, auto_bootstrap=False, debug=False, initial_token=None, add_node=True, is_seed=True, data_center=None, rack=None) -> ScyllaNode:
+        ipformat = self.get_ipformat()  # noqa: F841
```
Comment (fruch) [inline ccmlib/cluster.py:335]: What is the point of introducing noqa ? If we don't introduce the linter as well ?

File: ccmlib/cluster.py [hunk 4]
```diff
@@ -365,7 +368,7 @@ def get_storage_interface(self, nodeid):
         return (self.get_node_ip(nodeid), 7000)
 
     def get_node_jmx_port(self, nodeid):
-        return 7000 + nodeid * 100 + self.id;
```
Comment (fruch) [inline ccmlib/cluster.py:368]: Shame to get rid of ANSI C, isn't it ? 😜

File: ccmlib/cluster.py [hunk 5]
```diff
@@ -421,7 +424,7 @@ def remove_dir_with_retry(self, path):
                 try:
                     common.rmdirs(path)
                     removed = True
-                except:
+                except Exception:
```
Comment (fruch) [inline ccmlib/cluster.py:427]: Linter should be angry about this, isn't it ? too broad exception ?

I think in this case we can set a specific set of possible exceptions

**Review by fruch — APPROVED** [review]
LGTM

**Pull request merged**